source/assets/js/playground/utils.ts

```diff
@@ -0,0 +1,68 @@
+/* eslint-disable node/no-extraneous-import */
+import {Diagnostic} from '@codemirror/lint';
+import {Exception, OutputStyle, Syntax} from 'sass';
+
+import {ConsoleLog} from './console-utils';
+
+export type PlaygroundState = {
+  inputFormat: Syntax;
+  outputFormat: OutputStyle;
+  inputValue: string;
+  compilerHasError: boolean;
+  debugOutput: ConsoleLog[];
+};
+
+// State is persisted to the URL's hash format in the following format:
+// [inputFormat, outputFormat, ...inputValue] = hash;
+// inputFormat: 0=indented 1=scss
+// outputFormat: 0=compressed 1=expanded
+export function stateToBase64(state: PlaygroundState): string {
+  const inputFormatChar = state.inputFormat === 'scss' ? 1 : 0;
+  const outputFormatChar = state.outputFormat === 'expanded' ? 1 : 0;
+  const persistedState = `${inputFormatChar}${outputFormatChar}${state.inputValue}`;
+  return btoa(encodeURIComponent(persistedState));
+}
+
+export function base64ToState(string: string): Partial<PlaygroundState> {
+  const state: Partial<PlaygroundState> = {};
+  let decoded;
+  try {
+    // Remove hash
+    decoded = decodeURIComponent(atob(string.slice(1)));
```

Comment:
It's a little surprising to me that `stateToBase64` doesn't add a hash but `base64ToState` does remove it. I'd expect them to be pure inverses of one another.

Reply:
Good point, that bugged me as well. `location.hash` always returns `#` but doesn't require it on set, but it's cleaner to make these pure inverses. Fixed in https://github.com/sass/sass-site/pull/754/commits/94a2af06f86239d3e20009562770dfcc4186da26